syntax/text/parser/src/Luna/Syntax/Text/Pretty/Pretty.hs

```diff
@@ -201,64 +202,91 @@ instance ( MonadIO m -- DEBUG ONLY
+        Lam       arg body          -> named (notSpaced lamName) . atom .: (<>) <$> subgenBody arg <*> smartBlock body
+        LeftSection  op a           -> unnamed . atom . parensed .:      (<+>) <$> subgenBody op  <*> subgenBody a
+        RightSection op a           -> unnamed . atom . parensed .: flip (<+>) <$> subgenBody op  <*> subgenBody a
+        Marked       m a            -> unnamed . atom .: (<>) <$> subgenBody m   <*> subgenBody a
+        Marker         a            -> return . unnamed . atom $ convert markerBeginChar <> convert (show a) <> convert markerEndChar
+        ASGRootedFunction  n _      -> unnamed . atom . (\n' -> "<function '" <> n' <> "'>") <$> subgenBody n
+        ASGFunction  n as body      -> unnamed . atom .:. (\n' as' body' -> "def" <+> n' <> arglist as' <> body') <$> subgenBody n <*> mapM subgenBody as <*> smartBlock body
+        FunctionSig  n tp           -> unnamed . atom .: (\n' tp' -> "def" <+> n' <+> typedName <+> tp') <$> subgenBody n <*> subgenBody tp
+        Match        a cs           -> unnamed . atom .: (\expr body -> "case" <+> expr <+> "of" </> indented (block $ foldl (</>) mempty body)) <$> subgenBody a <*> mapM subgenBody cs
+        ClsASG _ n as cs ds         -> unnamed . atom .:. go <$> mapM subgenBody as <*> mapM subgenBody cs <*> mapM subgenBody ds where
+                                           go args conss decls = "class" <+> convert n <> arglist args <> body where
+                                               body      = if_ (not . null $ cs <> ds) $ ":" </> bodyBlock
+                                               bodyBlock = indented (block $ foldl (</>) mempty $ conss <> decls)
+
+        FieldASG mn a               -> unnamed . atom . (\tp -> if null mn then tp else intercalate space (convert <$> mn) <> Doc.spaced typedName <> tp) <$> subgenBody a
+        Invalid t                   -> return . named (spaced appName) . atom $ "Invalid" <+> convert (show t)
+        UnresolvedImportHub is      -> unnamed . atom . foldl (</>) mempty <$> mapM subgenBody is
+        UnresolvedImport i t        -> unnamed . atom . (\src -> "import " <> src <> tgts) <$> subgenBody i where
+                                       tgts = case t of Import.Everything -> ""
+                                                        Import.Listed ns  -> ": " <> intercalate " " (convert <$> ns)
+        UnresolvedImportSrc i       -> return . unnamed . atom $ case i of
+            Import.World            -> "#World#"
+            Import.Absolute ss      -> convertVia @P.String ss
+            Import.Relative ss      -> "." <> convertVia @P.String ss
+
+        ForeignImportList language imports ->
+            unnamed . atom . (\imps -> "foreign import"
+                <+> convert language
+                <>  lamName
+                </> indented (block $ foldl (</>) mempty imps))
+            <$> mapM subgenBody imports
```

Comment:
Again, such code could be formatted to much more concise and arguable more readable form:

```
ForeignImportList lang imps -> unnamed . atom . cons <$> mapM subgenBody imps
    where cons imps = "foreign import" <+> convert lang <> lamName
                  </> indented (block $ foldl (</>) mempty imps))
```